File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.diameter = 0
        
        def maxDepth(node):
            if not node:
                return 0
            leftmax = maxDepth(node.left)
            rightmax = maxDepth(node.right)
            self.diameter = max(self.diameter, leftmax + rightmax)
            return 1 + max(leftmax, rightmax)
        
        maxDepth(root)
        return self.diameter

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int diameter = 0;
    
    public int diameterOfBinaryTree(TreeNode root) {
        maxDepth(root);
        return diameter;
    }
    
    private int maxDepth(TreeNode root) {
        if (root == null) return 0;
        int leftmax = maxDepth(root.left);
        int rightmax = maxDepth(root.right);
        diameter = Math.max(diameter, leftmax + rightmax);
        return 1 + Math.max(leftmax, rightmax);
    }
}